### DISCUSSION TITLE
(nextjs) - useHydrateAtoms return previous value on dynamic route

### DISCUSSION THREAD

**Discussion by zecka** [pmndrs/jotai#2252] [Bug report]
## Summary
I use the `useHydrateAtoms` hooks on a dynamic route. 
On server side atom initialize with `useHydrateAtoms` use value from the previous render:

Example:
- i have following route /pages/[slug.tsx]
- Go to localhost:3000/test -> Everything ok
- Go to localhost:3000/test2 -> Error, server return value from /test

![GIF 13-11-2023 17-11-55](https://github.com/pmndrs/jotai/assets/18116930/0f00aa95-de33-421a-9d9d-200dcb24793d)

```tsx
//pages/[slug].tsx
export default function Slug({slug}: SlugPageProps) {
  useHydrateAtoms([
    [slugAtom, slug],
  ])
  const slugValue = useAtomValue(slugAtom)
    return (
        <div>slugValue: { slugValue}</div>
  )
}


export const getStaticProps: GetStaticProps<SlugPageProps> = async ({ params }) => {
const slug = params?.slug as string ?? '';
  return {
    revalidate: 60,
    props: {
        slug,
    },
  };
};
```

## Link to reproduction
Here is a reproductible example:
https://github.com/zeckaissue/jotai-useHydrateAtoms-next-bug
https://codesandbox.io/p/github/zeckaissue/jotai-useHydrateAtoms-next-bug

The only solution i found to fix that is to use `dangerouslyForceHydrate`, but not sure is really recommanded:
```tsx
useHydrateAtoms(
    [
        [slugAtom, slug],
    ],
    { dangerouslyForceHydrate: typeof window === 'undefined' }
)
```
(Note that maybe this is a normal behaviour and not a bug...)

**Comment by dai-shi** [comment]
If `dangerouslyForceHydrate` works for your case, it's exactly how it's designed.
useHydrateAtoms by default only set values of atoms that have never been initialized. Otherwise, it can violate React rendering rule (= being idempotent).

Using `Provider` under the route might help, because it will create a fresh store for different routes.

**Reply by zecka** [reply]
Okay got it. 

But as you say Provider can help if the issue is on the different route, so not in this case right ?

Typically i got this issue on multi-locale website (with next-translate). I have some atom data that contain localized string.

BTW `dangerouslyForceHydrate` works, I just felt like I was doing something wrong by using it because of the word "`dangerously`".

Thank you so much for your answer, your reactivity on the questions in your repositories jotai and zustand is impressive.

**Reply by dai-shi** [reply]
Yes, "dangerously" should be avoided unless it's really something intended.

Have you tried useEffect? It will show the previous value on the first render, and then updated afterwards.

> on multi-locale website

Not sure if it works, but adding `key={locale}` to Provider might do?

**Reply by zecka** [reply]
Yes I tried useEffect and as you say the value is well updated on client side, but the SSR still have wrong value.
I will check check with `key={local}` on Provider and put my feedback here !
